onnxruntime/core/providers/cpu/math/einsum_utils/einsum_auxiliary_ops.cc

```diff
@@ -251,6 +251,28 @@ std::unique_ptr<Tensor> Diagonal(const Tensor& input, int64_t dim_1, int64_t dim
 
 }  // namespace DeviceHelpers
 
+// This helps decide if we need to reshape instead of transposing.
+bool IsReshapeRequired(const std::vector<int64_t>& input_dims, const std::vector<size_t>& permutation) {
+  ORT_ENFORCE(input_dims.size() == permutation.size(), "The rank of the input must match permutation size for Transpose/Reshape.");
+
+  // No transpose required for scalars
+  if (input_dims.size() == 0) {
+    return false;
+  }
+
+  // A transposition only moving single dimension is equivalent to a reshape.
```

Comment:
Just a thought - Can this stub actually be part of Transpose itself (i.e.) once we recognize that the transpose request is just a reshape request within the Transpose op, can we just do a memcpy of the input buffer into the output buffer (without invoking other fancier components of Transpose) and stamp the copied output buffer with the new shape and return that Tensor back from Transpose? Does this special-casing live in Transpose currently ? If it does not, can we check how much gains that single enhancement can give ?

My concern is with re-sharing a piece of memory across multiple Tensors and synchronizing the life-cycle of the memory so that it is valid till the last Tensor using it is valid. This is going to need some work in the Tensor class...

Reply:
I already did that. In the first PR, operator Transpose detects the case where a Transpose is a just Reshape. Then the data is copied in one block. In this PR, I wanted to even avoid the allocation: the gain I measured 300ms out of 3700ms is the time spent in allocation + copy of Reshaped/Transposed Tensor. I cannot remove the allocation in operator Transpose because the reshaped Tensor cannot exist without the Tensor owning the buffer and there is no way to make sure of that. In operator Einstein, the tranposed are intermediate tensors, their scope is known. The reshaped tensor disappears before or at the same time as the tensor owning the buffer.

Reply:
A more general comment, operator Transpose or Reshape could probably reuse the buffer of the input Tensor if this operator is the only consumer of this tensor.

Reply:
> A more general comment, operator Transpose or Reshape could probably reuse the buffer of the input Tensor if this operator is the only consumer of this tensor.

If it's able to be determined upfront that the buffer can be re-used an optimizer could replace the Transpose with a Reshape. That requires the perms and the input shape to be known. 

I don't think it's possible to re-use otherwise if there's an allocation plan, as an input buffer which was expected to become unused after running the Transpose would still be in use.

Reshape re-uses the buffer if possible as the input/output is marked as Alias in the kernel registration.